### ISSUE TITLE
validator-engine memory leak (when validating)

### ISSUE THREAD

**Issue by Skydev0h** [ton-blockchain/ton#235]
When validating, validator-engine memory usage raises gradually over time until the VE gets reaped by OOM killer.

I have tried to collect some evidence with jemalloc (kudos to Arseny for the idea).
As for now I suspect that some ValidatorSessionDescriptionImpl are unneccessarily immortal and get collected over time.

Another thing that caught my eye is that ValidatorSessionImpl has a start_up method but has no tear_down method.
Should be noted that half of those Descs are spawned in ValidatorSessionImpl::start_up calling ValidatorSessionState::move_to_persistent, another half are created by create_actor.

Some food for thinking (my attempts at building some deltas):
![app-delta](https://user-images.githubusercontent.com/5331782/73294590-d2183a80-420e-11ea-825a-1a008f998aad.gif)
---
![app-delta-2](https://user-images.githubusercontent.com/5331782/73294591-d2183a80-420e-11ea-9b1e-fea09381f45a.gif)
---
![app-delta-3](https://user-images.githubusercontent.com/5331782/73294592-d2183a80-420e-11ea-83a4-dfbbe57a395c.gif)

**Comment by Skydev0h** [comment]
There is actually another thing that is gradually raising in space usage, on the very bottom of the graph. It may be the culprit (I dont see much raise in ValidatorSessionDescriptionImpl, it is still 2048+2048, while the bottom one slowly raises.
![app-delta-4](https://user-images.githubusercontent.com/5331782/73298500-11965500-4216-11ea-8d7e-b10a788d8b4e.gif)

**Comment by Skydev0h** [comment]
Last prof before OOM reaper indicates that ValidatorSessionDescriptionImpl totally still uses same 4096 MB of memory, while the box in bottom (rocksdb::BlockFetcher::PrepareBufferForBlockFromFile) is constantly rising, slowly but steadily. Judging from the pace of normal memory rise when used, I think that it may be the culprit.

![app-delta-5](https://user-images.githubusercontent.com/5331782/73316362-52ed2b80-423b-11ea-928f-39f20e95760b.gif)

**Comment by Skydev0h** [comment]
Focusing on that little box those are changes over time:
![image](https://user-images.githubusercontent.com/5331782/73316576-f3dbe680-423b-11ea-8e8c-27abfe26c293.png)
![image](https://user-images.githubusercontent.com/5331782/73316580-f63e4080-423b-11ea-9362-263389995fe7.png)
![image](https://user-images.githubusercontent.com/5331782/73316583-f8a09a80-423b-11ea-9146-18ff6e47ac87.png)
![image](https://user-images.githubusercontent.com/5331782/73316589-fc342180-423b-11ea-8869-fd4b3c2b07ac.png)
Looks like a typical memory leak.

**Comment by Skydev0h** [comment]
Some memory profile of validator over night.
Despite the sawtooth, lower edge of it steadily rises over time.
And that slow and steady rise has nearly the same pace as rise of that last block I found.

![mem-graph - Copy](https://user-images.githubusercontent.com/5331782/73334060-489b5380-4274-11ea-9e4f-4d747fa5a323.png)

Approximately 200 MB over 14500‬ seconds (~4 hours), that is 50 MB per hour.
That would take, for example, just 13.6 days to fill 16 GB of RAM.
That is not counting the sudden spikes that are not reclaimed later.

**Comment by Skydev0h** [comment]
It may be possible that I have found a fix, but still need several days for testing. May be not.
But looks a little more managable with some more shelves.
Before:
![image](https://user-images.githubusercontent.com/5331782/73490246-c5cde200-43b4-11ea-846f-f89268485f6e.png)
After:
![image](https://user-images.githubusercontent.com/5331782/73490224-bbabe380-43b4-11ea-9a76-35a6042558d0.png)

**Comment by hortonelectric** [comment]
@Skydev0h Glad to see someone tackling this. The solution for now is to use systemctl daemon to restart the validator engine whenever it crashes due to this leak. It's great for testing failure of the core system :)

You're talking about 13.6 days but I've had very mixed results. Sometimes it takes an hour, sometimes a week. It does not correlate with processing incoming messages, you can run a 'master node' (with its own zerostate) and just watch it eat itself.

**Comment by Skydev0h** [comment]
@ton-blockchain I seem to have stabilized to some extent at least one possible leak factor. Should require more testing, but graph looks more stable now (at least it keeps stable memory usage for more time). May be a first step but not the last.
![image](https://user-images.githubusercontent.com/5331782/73590864-5bba5780-44f0-11ea-8e99-469f027c0d43.png)
Please notice how it stopped constantly raising memory after cache (the 1GB NewLRUCache(1 << 30) maybe?) got filled up, and now possibly leaks for some another reason (I think that those sudden +1GB and +2GB spikes are arena allocations for ValidatorSessionDescriptionImpl mentioned earlier)
I am not even making a PR because it is one line change.
![image](https://user-images.githubusercontent.com/5331782/73590844-f36b7600-44ef-11ea-86dc-84bcc78a5824.png)
This may _theoretically_ degrade perfomance under some circumstances, but otherwise index and filter blocks are stored in never-evicted heap and may limitlessly consume memory.
It may be reasonable to set _cache_index_and_filter_blocks_with_high_priority_, but I have not yet observed perfomance problems and increased amount of SLOWs in validator logs.
![image](https://user-images.githubusercontent.com/5331782/73590861-4e04d200-44f0-11ea-9e1f-8e5711811bda.png)

**Comment by Skydev0h** [comment]
@hortonelectric does the masternode do validation tasks?
Or just a simple full node? I did not observe memory leaks for simple full node.

**Comment by hortonelectric** [comment]
Yes i am talking about a validator... In my regtest setup i use a single
node that does everything hence "master"

**Comment by jetaweb** [comment]
Got the same problem, memory leak by waves.
